### PR TITLE
[feat] Add benchmark tools

### PR DESCRIPTION
**Pull request by reciprocated**
This PR adds tools to cross-reference example runs from different branches.

As an example, the following command will `git clone` [add-benchmark-tools](https://github.com/CarperAI/trlx/tree/add-benchmark-tools) branch and run a set of examples on it, storing them under `trlx-references` project name under a private username:
```bash
python -m trlx.reference add-benchmark-tools
```
Also it will run the same set of examples on `main` by default, unless they were already present on the `trlx-references` project name, and make a w&b report comparing the two sets. Example of a report: https://wandb.ai/sorry/trlx-references/reports/add-benchmark-tools-v-main--VmlldzozOTAwMTQy

Presency of examples' runs is determined by a `tags` config field which contains hash of all files (excluding markdown). If runs are made on top of some PR, and after the PR is merged into the `main`, the total hash of files will not change (unlike git commit hash), so that runs won't have to be remade again.

By default runs are referenced against CarperAI/trlx's `main`, but optionally they can be referenced against any branch:
```bash
python -m trlx.reference xu-song/trlx:fix1 --against CarperAI/trlx:convert-examples-configs
```

### CODE REVIEW

File: scripts/benchmark.sh
Comment (cat-state) [inline scripts/benchmark.sh:42]: is there some way to do this without mutating the environment? since this can break stuff you had installed depending on one version of torch.
Also would probably be good to pin a torch version and tag it? Or otherwise save the `pip freeze` output somewhere in the run's metadata.

Comment (cat-state) [reply scripts/benchmark.sh:42]: It also failed to install for me with 
```
  Using cached https://files.pythonhosted.org/packages/bc/c3/f068337a370801f372f2f8f6bad74a5c140f6fda3d9de154052708dd3c65/Jinja2-3.1.2-py3-none-any.whl
Collecting triton==2.0.0; platform_system == "Linux" and platform_machine == "x86_64" (from torch)
  ERROR: Could not find a version that satisfies the requirement triton==2.0.0; platform_system == "Linux" and platform_machine == "x86_64" (from torch) (from versions: 0.4.1, 0.4.2)                                                 
ERROR: No matching distribution found for triton==2.0.0; platform_system == "Linux" and platform_machine == "x86_64" (from torch)                         
```

**Review by cat-state — COMMENTED** [review]
Thanks for this @reciprocated ! This looks good to me mostly except for a few suggestions, however when I tried to run it I get 
```
  File "/mnt/nvme/home/uwu/conda/nemo-113/lib/python3.8/site-packages/wandb/sdk/wandb_require_helpers.py", line 37, in __init__
    self._check_if_requirements_met()
  File "/mnt/nvme/home/uwu/conda/nemo-113/lib/python3.8/site-packages/wandb/sdk/wandb_require_helpers.py", line 48, in _check_if_requirements_met
    raise Exception(
Exception: You must explicitly enable this feature with `wandb.require("report-editing:v0)"
```

File: 0001-feat-configs-add-tags-config-option.patch
Comment (reciprocated) [reply 0001-feat-configs-add-tags-config-option.patch:1]: I plan to delete this altogether right before this pr is merged, since it just patches other branches with a subset of changes this pr introduces

Comment (cat-state) [reply 0001-feat-configs-add-tags-config-option.patch:1]: got it, that makes sense

File: scripts/benchmark.sh
Comment (reciprocated) [reply scripts/benchmark.sh:45]: It's already logged under "System Hardware" in w&b as `GPU type:	NVIDIA A100-SXM4-80GB`, or do you want to see it specifically as a tag?

File: scripts/benchmark.sh
Comment (reciprocated) [reply scripts/benchmark.sh:42]: This setup directly reflects installation instructions from the https://github.com/CarperAI/trlx#readme. However I wouldn't mind pinning every dependency down both here and there

Also this is odd because I can install `triton==2.0.0` on `uname -i == x86_64`, moreover we probably use the same hardware here

File: scripts/benchmark.sh
Comment (reciprocated) [reply scripts/benchmark.sh:55]: nothing noticable https://wandb.ai/sorry/trlx/reports/Timing-difference--VmlldzozODA4MTA4

**Comment by reciprocated** [comment]
@cat-state you must be using some older `wandb` version, could you please try it again after upgrading to `wandb==0.14.0`?

**Review by cat-state — CHANGES_REQUESTED** [review]
Thanks, i was able to run it after updating wandb. I think this would be good to merge after adding the pip upgrade and the `$dir` check. I also still get 
```
wandb.errors.CommError: It appears that you do not have permission to access the requested resource. Please reach out to the project owner to grant you access. If you have the correct permissions, verify that there are no issues with your networking setup.(Error 403: Forbidden)
```
even though I have a `trlx-references` project in my wandb.

**Review by cat-state — COMMENTED** [review]
Also add a link to the trlx-references wandb in the readme

File: scripts/benchmark.sh
Comment (reciprocated) [reply scripts/benchmark.sh:24]: i don't think there is, but an additional check would hurt either

**Review by cat-state — APPROVED** [review]

**Comment by cat-state** [comment]
Got it to work after changing my default entity from carperai to myself, thank you!

**Pull request merged**
